syntax = "proto3"; // 版本
package fixbug; // 声明代码所属的包 类似namespace

option cc_generic_services = true; // 定义下面的选项，表示生成service服务类和rpc方法描述

// 数据     列表
// 定义登录消息类型 
message LoginRequest {
    bytes name = 1;
    bytes pwd = 2;
}

message ResultCode {
    int32 errcode = 1;
    bytes errmsg = 2;
}

// 定义登录响应消息类型
message LoginResponse {
    ResultCode result = 1; // 嵌套message
    bool success = 2;
}

message GetFriendListsRequest {
    uint32 userid = 1;
}

message User {
    bytes name = 1;
    uint32 age = 2;
    enum Sex {
        MAN = 0;
        WOMAN = 1;
    }
    Sex sex = 3;
}   

message GetFriendListsResponse {
    ResultCode result = 1;
    repeated User friendlists = 2; // 数组
}

// 在protobuf中定义一个RPC服务接口，通过protoc可以生成一个抽象接口UserServiceRpc以及一个相应的存根实现（Stub）
// 存根将所有Rpc方法的调用指向RpcChannel，它是一个抽象接口，必须在Rpc系统中对RpcChannel的纯虚函数CallMethod()进行实现
// 服务提供方通过服务注册来生成一个ServiceRpc，服务调用方通过Stub类来实现对Rpc方法的调用
service UserServiceRpc {
    rpc Login(LoginRequest) returns(LoginResponse);
    rpc GetFriendList(GetFriendListsRequest) returns(GetFriendListsResponse);
}